Tabs: sync browser focus to selected tab in controlled mode

chad1008 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
In controlled mode, make browser focus follow any changes in tab selection that occur while a rendered tab currently has browser focus.

## Why?
In some cases, it's possible for focus to be passed to the selected tab, only to have that selection change immediately afterwards. This could leave keyboard navigation and/or assistive tech in a confusing state where the focused tab does not reflect the content actually being displayed.

Real life example in the editor when [tabbing to the sidebar without having a block actively selected](https://github.com/WordPress/gutenberg/pull/55360#issuecomment-1828420884).

## How?
The main `Tabs` component now wraps its content provider in a div that we can attach a ref to. Using that ref, we can identify what DOM element currently has focus in the browser. Comparing that element's id to our tabs ids, we can then then make sure that the focus remains on the selected tab at the right times (specifically, only if a tab was actively focused when the change occurred).

I wasn't able to use Ariakit's `activeId` value for this, because it appears to already follow the selected tab, which can leave it out of sync with browser focus if conditions are right as in the example above

## Testing Instructions

<details>
<summary>
Apply the following diff to perform this test in Storybook:</summary>

```diff
diff --git a/packages/components/src/tabs/stories/index.story.tsx b/packages/components/src/tabs/stories/index.story.tsx
index ce8c8324ed..2a724c0171 100644
--- a/packages/components/src/tabs/stories/index.story.tsx
+++ b/packages/components/src/tabs/stories/index.story.tsx
@@ -273,17 +273,29 @@ const ControlledModeTemplate: StoryFn< typeof Tabs > = ( props ) => {
 					<DropdownMenu
 						controls={ [
 							{
-								onClick: () => setSelectedTabId( 'tab1' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab1' ),
+										3000
+									),
 								title: 'Tab 1',
 								isActive: selectedTabId === 'tab1',
 							},
 							{
-								onClick: () => setSelectedTabId( 'tab2' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab2' ),
+										3000
+									),
 								title: 'Tab 2',
 								isActive: selectedTabId === 'tab2',
 							},
 							{
-								onClick: () => setSelectedTabId( 'tab3' ),
+								onClick: () =>
+									setTimeout(
+										() => setSelectedTabId( 'tab3' ),
+										3000
+									),
 								title: 'Tab 3',
 								isActive: selectedTabId === 'tab3',
 							},

```

</details>
The above diff adds a delay to the controlled-mode selection changes. This will allow you to simulate the tab changing on you while it's already focused.

1. Launch Storybook and open the "Controlled Mode" story
2. Using the dropdown, choose a tab to switch to
3. Before it updates (you have three seconds) click somewhere on the canvas, and press [Tab] to focus the currently selected tab.
4. Note that when the controlled selection takes effect, the newly selected tab receives browser focus.
5. Repeat this test, but skip step three. Don't force the focus onto a tab.
6. Confirm that the focus doesn't move when the selection kicks in (in this case, it stays on the dropdown toggle)

## Notes

In most implementations, if a tab has focus it will automatically be the selected tab, because by default focusing a tab selects it. The exception to this is when `selectOnMove = false`. This means that in controlled mode, with that prop set to `false` it is possible for a user to be navigating tabs with their arrow keys at the same time that a controlling component forces a selection change. I expect that to be pretty rare, and this change will mostly take effect during race conditions like the one in the example above. I don't anticipate it happening often while someone's casually navigating a bit of UI. It is technically possible though, and with this PR in effect that would move the focus to the newly selected tab. I went back and forth on how to approach that.

On the one hand, we could limit this change to only fire when `selectOnMove = true`, or fine-tune it a bit to only take effect if the selected tab specifically was focused when the controlled selection change took effect.

On the other hand, I thought the change in focus might be a good way to notify assistive tech that something had just happened, rather than letting it happen silently in the background with no warning. Very much open to feedback on this point! cc @andrewhayward in case you have thoughts.

chad1008 committed:
link browser focus to selected tab changes in controlled mode

ciampo commented on packages/components/src/tabs/index.tsx:
```diff
@@ -42,10 +42,20 @@ function Tabs( {
 	} );
 
-	const isControlled = selectedTabId !== undefined;
+	const isControlled = useMemo(
+		() => selectedTabId !== undefined,
+		[ selectedTabId ]
+	);
```
What's the reason for memoizing this variable?

chad1008 replied:
I think that was leftover from something else I'd been experimenting with, thanks for spotting it.

ciampo commented on packages/components/src/tabs/index.tsx:
```diff
@@ -154,6 +164,22 @@ function Tabs( {
+
+		if (
+			tabsHasFocus &&
+			tabsRef.current?.ownerDocument.activeElement !== null &&
+			selectedId !== tabsRef.current?.ownerDocument.activeElement.id
```
It looks like the `tabHasFocus` variable already includes a check about the `activeElement` not being falsy (ie. also not `null`) — is there a reason why the check is repeated here?

If not, we could rewrite this check as

```suggestion
			selectedId !== tabsRef.current?.ownerDocument.activeElement?.id
```

ciampo commented on packages/components/src/tabs/index.tsx:
```diff
@@ -154,6 +164,22 @@ function Tabs( {
 	] );
 
+	// In controlled mode, make sure browser focus follows the selected tab if
+	// the selection is changed while a tab is already being focused.
```
I feel like we shouldn't apply this behaviour when `selectOnMove` prop is set to `false`

chad1008 replied:
That works for me, I really couldn't decide how I felt on that. Added a `selectOnMove` check.

ciampo commented on packages/components/src/tabs/index.tsx:
```diff
@@ -163,9 +189,11 @@ function Tabs( {
-		<TabsContext.Provider value={ contextValue }>
-			{ children }
-		</TabsContext.Provider>
+		<div ref={ tabsRef }>
```
I wonder if we should find an alternative way to know whether any tabs are currently focused that doesn't require us to render a wrapper `div`

diegohaz replied:
You could check `item.element`:

```js
const tabsHasFocus = items.some( ( item ) => {
  const activeElement = item.element?.ownerDocument.activeElement;
  return activeElement && activeElement === item.element;
} );
```

However, I think that should be done inside `useEffect` rather than on render.

ciampo replied:
Yeah, agreed. Thank you!

chad1008 replied:
Thank you both! Using `item.element` was a great idea. I had to tweak the approach slightly, because I needed to rescope `activeElement` so I could access it again in a different scope, but the changes in ea1e3233e0 have this working without adding a `div` solely to carry a ref around.

ciampo reviewed (commented):
Thank you for working on this!

It would be great if we had unit tests, so that we could also test for alternative solution much quicker!

I would also love to hear from @diegohaz  in case he had any thoughts on alternative solutions

chad1008 committed:
remove unnecessary memoization

chad1008 committed:
remove duplicate condition

chad1008 committed:
dont apply when selectOnMove is false

andrewhayward commented:
> On the one hand, we could limit this change to only fire when `selectOnMove = true`, or fine-tune it a bit to only take effect if the selected tab specifically was focused when the controlled selection change took effect.
> 
> On the other hand, I thought the change in focus might be a good way to notify assistive tech that something had just happened, rather than letting it happen silently in the background with no warning.

Tricky one. Refocusing can be very disorienting, particularly when you don't know why it happened, so [should be avoided](https://www.w3.org/WAI/WCAG21/Understanding/on-focus.html). But at the same time, being out of sync with a control can also be confusing, particularly when focus is tied to selection state.

I'd err on not switching focus over staying in sync, but if we can prevent that from happening too then so much the better. [Looks like you're already accounting for `selectOnMove`](https://github.com/WordPress/gutenberg/pull/56658/commits/308a03618117996bb730dd6a858ee68fe76ff2e7)? Does this PR now resolve [the issue in #55360](https://github.com/WordPress/gutenberg/pull/55360#issuecomment-1828420884)?

chad1008 committed:
use `items` to find active element, instead of a ref

chad1008 commented:
> I'd err on not switching focus over staying in sync, but if we can prevent that from happening too then so much the better. [Looks like you're already accounting for selectOnMove](https://github.com/WordPress/gutenberg/pull/56658/commits/308a03618117996bb730dd6a858ee68fe76ff2e7)? Does this PR now resolve https://github.com/WordPress/gutenberg/pull/55360#issuecomment-1828420884?

Perfect, thank you for the input! And yes, this PR should address that issue.

chad1008 committed:
CHANGELOG

ciampo commented:
The code is looking good! Could you also add a unit test for the desired behaviour (both controlled/uncontrolled, selectOnMove true/false) ?

chad1008 committed:
add `selectOnMove` unit tests

chad1008 commented:
> Could you also add a unit test for the desired behaviour (both controlled/uncontrolled, selectOnMove true/false) ?

I've added the new tests, but they're currently flakey. It doesn't look like the tests themselves, as they pass consistently when run individually. It's when I run them together that they fail.

There are two new uncontrolled mode tests, and I only see the second one occasionally being flaky, not the first. There are four new controlled mode tests, and the flakiness there also appears isolated to the second, third, and/or fourth but not the first.

All of that to say, it really feels like something is leaking between tests but I can't tell what, and I haven't yet figured out how to prevent it.

One other note: there isn't really a way to test the new behavior of this PR in uncontrolled mode, due to the nature of the changes, so the uncontrolled tests are really only testing that `selectOnMove` works properly, but not the new stuff. Still good tests to have though!

I'm going to switch tasks for a bit, but I'll circle back to continue investigating what's leaking between these tests later.

chad1008 commented:
Update: moving from `userEvent` to `ariakit/test` utils appears to resolve the timing/leaking issues I was seeing in unit tests.  (h/t @ciampo for the suggestion).

I'm moving the existing tests to `ariakit/test` in #56835, and will update the new tests on this PR once it has been rebased.

chad1008 committed:
convert new tests to ariakit/test

chad1008 commented:
The `ariakit/test` changes are all in place, and the tests are passing consistently for me locally 🎉

chad1008 commented:
CI checks are finally passing! One of them was oddly stubborn, but passes consistently locally, and is now fine here as well.

ciampo commented on packages/components/src/tabs/test/index.tsx:
```diff
@@ -795,6 +799,57 @@ describe( 'Tabs', () => {
+				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
+				expect( await getSelectedTab() ).toHaveFocus();
+			} );
+		} );
+
+		describe( 'When `selectOnMove` is `false`', () => {
+			it( 'should apply focus without automatically changing the selected tab', async () => {
+				render(
+					<UncontrolledTabs tabs={ TABS } selectOnMove={ false } />
+				);
+
+				// Tab should focus the currently selected tab, which is Alpha.
+				await press.Tab();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
+				expect( await getSelectedTab() ).toHaveFocus();
+
+				// Arrow key should move focus but not automatically change the selected tab.
+				await press.ArrowRight();
+				expect(
+					screen.getByRole( 'tab', { name: 'Beta' } )
+				).toHaveFocus();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Alpha' );
+
+				// Pressing the spacebar should select the focused tab.
+				await press.Space();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
+
+				// Arrow key should move focus but not automatically change the selected tab.
+				await press.ArrowRight();
+				expect(
+					screen.getByRole( 'tab', { name: 'Gamma' } )
+				).toHaveFocus();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Beta' );
+
+				// Pressing the enter/return should select the focused tab.
+				await press.Enter();
+				expect( await getSelectedTab() ).toHaveTextContent( 'Gamma' );
+			} );
+		} );
+
```
These tests are a bit redundant:

- `selectOnMove={false}` with the uncontrolled version is already tested in the "'switches to manual tab activation when the `selectOnMove` prop is set to `false`" test
- `selectOnMove={true}`  with the uncontrolled version is kind of the default behaviour that is tested already in every test inside the "Tab Activation" section

chad1008 replied:
Shoot, I'd noticed the redundancy of that first one but totally forgot about it when pushing the last set of changes.

You make a good point on the second one being covered in multiple other tests as well. Both are removed.

ciampo reviewed (commented):
Not something for this PR but rather for a potential follow-up:

I may be forgetting some context about past decisions, but is there a reason why the "Accessibility and semantics", "Focus Behavior", "Tab Attributes", and "Tab Activation" sub-sections in the tests are only testing the uncontrolled version of the component? And why do we have then dedicated "Uncontrolled mode" and "Controlled mode" sections?

I feel like a better structure could be to:
- have a top-level `describe.each` which makes sure that we're testing both uncontrolled and controlled version of the components
- for controlled-only and uncontrolled-only, we could have separate sections (or subsections)

chad1008 committed:
remove redundant tests

chad1008 commented:
> I may be forgetting some context about past decisions, but is there a reason why the "Accessibility and semantics", "Focus Behavior", "Tab Attributes", and "Tab Activation" sub-sections in the tests are only testing the uncontrolled version of the component? And why do we have then dedicated "Uncontrolled mode" and "Controlled mode" sections?

If I recall correctly, that structure grew out of how we implemented Controlled Mode initially, [with a focus on how tab activation would differ (or not) between the two modes](https://github.com/WordPress/gutenberg/pull/53960#issuecomment-1716259711). When I added the "Controlled" and "Uncontrolled" sections originally, it was to test those differences in behavior between the two modes. The tests in those first sections you've mentioned are things that shouldn't be impacted by the behavioral changes between the two modes.

> I feel like a better structure could be to:
> - have a top-level describe.each which makes sure that we're testing both uncontrolled and controlled version of the components
> - for controlled-only and uncontrolled-only, we could have separate sections (or subsections)

A problem with a top-level `describe.each` (but please correct me if there's a good way around this) is that a fair number of tests will be different in controlled vs uncontrolled, so the assertions would change. For example in the table linked above, an assertion for what the selected tab should be when there is no `initialTabId` provided wouldn't be the same for both runs of the `.each`. We'd probably end up with a lot of conditional assertions.

All of that said, I'm more than happy to look into reworking the testing suite's structure! At the very least, we can probably run more of the tests through both modes, making _sure_ controlled/uncontrolled doesn't impact them in ways they shouldn't.

ciampo reviewed (approved):
🚀 

Regarding the structure of the test file, definitely not a priority right now anyway.
